sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -1148,6 +1165,12 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
         EqualNullSafe(e, expression(ctx.right))
       case SqlBaseParser.DISTINCT =>
         Not(EqualNullSafe(e, expression(ctx.right)))
+      case SqlBaseParser.ANY | SqlBaseParser.SOME =>
+        invertIfNotDefined(AnySubquery(
+            getValueExpressions(e),
+            getComparisonOperator(
+              ctx.comparisonOperator.getChild(0).asInstanceOf[TerminalNode].getSymbol.getType),
+            ListQuery(plan(ctx.query))))
```

Comment:
We cannot define an actual binary comp expr here?, e.g., 
```
case AnySubquery(cmp: BinaryComparison)

getComparison(left: Expression, right: Expression): BinaryComparison = { ... }

AnySubquery(getComparison(left = getValueExpression(e), right = ListQuery(plan(ctx.query))))

```

Reply:
We already have the similar logic, so can we reuse it?
https://github.com/apache/spark/blob/10539dd4e5fbac7cde0971b55320878156ff4e94/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L1073

Reply:
https://github.com/apache/spark/blob/10539dd4e5fbac7cde0971b55320878156ff4e94/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L1079
Here， `SqlBaseParser.NEQ` and `SqlBaseParser.NEQJ` will be parsed as Not(EqualTo(left, right)) which returns an `Expression` instead of `BinaryComparison`.
I have an idea to reuse it, but still need to re-match.
We can use `NotEqualTo(left, right)` in `getComparison(operator:Int, left: Expression, right:Expression)` but re-match it later.
```scala
def getComparison(operator: Int, left: Expression, right: Expression): BinaryComparison = operator match{
  ...
  case SqlBaseParser.NEQ | SqlBaseParser.NEQJ =>
        NotEqualTo(left, right)
  ...
}
```
```scala
def visitComparison(ctx: ComparisonContext): Expression = {
  ...
  getComparison(operator, left, right) match {
    case _: NotEqualTo =>
      Not(EqualTo(left, right))
    case _ =>
      _
  }
}
```
As for `AnySubquery`, we still need a  function to rewrite these expression(used in `RewritePredicateSubquery`).

```scala
def getComparisonExpression(cmp: BinaryComparison)(left: Expression, right: Expression)
  : Expression = cmp match{
  case _: EqualTo => EqualTo(left, right)
  ...
  case _: NotEqualTo => Not(EqualTo(left, right))
  ...
}
```
Do you have any better ideas?

Reply:
For now, IMO its ok to do so.
`getComparisonExpression` -> `getPredicate`?

Reply:
Hi, I found another problem during revision. 
If we use `AnySubquery(cmp: BinaryComparison)`, `left` inside the `cmp` is an `Expression` instead of `Seq[Expression]`, which can not be analyzed by the rule `ResolveReferences`(since it uses `mapProductIterator`). 
Therefore, I have to define it as 
```scala
case class AnySubquery(values: Seq[Expression], cmp: BinaryComparison, query: ListQuery)
```
and define another apply function as 
```scala
object AnySubqury {
  def apply(cmp: BinaryComparison): AnySubquery = ...
}
```
As a result, this becomes more and more unnatural.

I am wondering if it is necessary to restrict the type of `comparison` as `BinaryComparison`. If it is not, then we won't need `NotEqualTo()` any more and we can define `AnySubquery` as 
```scala
case class AnySubquery(values: Seq[Expression], comparison: (Expression, Expression) => Expression, query: ListQuery)
```
There is no need to rewrite `comparison` either.

Reply:
Another solution is that we can parse `x != ANY(query)` to `Not(AnySubquery(a, EqualTo, query))`, so that we can keep the `comparison` as type `(Expression, Expresson) => BinaryComparison`